src/page-storage/sameness.js

```diff
@@ -0,0 +1,68 @@
+import get from 'lodash/fp/get'
+import diffMatchPatch from 'diff-match-patch'
+
+export const Sameness = {
+    EXACTLY: 5,    // Perfect match, containing exactly the same data.
+    OSTENSIBLY: 4, // Semantically the same, differing only in e.g. whitespace.
+    MOSTLY: 3,     // Only minor visible edits, e.g. some typo corrections.
+    PARTLY: 2,     // Some content changed, e.g. parts were added or (re)moved.
+    HARDLY: 1,     // Whole page seems newly created, but with similar purpose.
+    UNRELATED: 0,  // Completely different topic, e.g. domain was transferred.
+    UNKNOWN: NaN,  // Not enough information to make a comparison.
+}
+
+// Get a rough similarity metric between strings (a number between 0 and 1).
+function stringSimilarity(text1, text2) {
+    const dmp = new diffMatchPatch()
+    dmp.Diff_Timeout = 0.1
+    const diff = dmp.diff_main(text1, text2)
+    dmp.diff_cleanupSemantic(diff)
+    const distance = dmp.diff_levenshtein(diff)
+    const maxDistance = Math.max(text1.length, text2.length)
+    const similarity = 1-distance/maxDistance
+    return similarity
+}
+
+// Tell how similar two strings are in a qualitative way.
+function textSameness(text1, text2) {
+    if (!text1 || !text2)
+        return Sameness.UNKNOWN
+
+    if (text1 === text2)
+        return Sameness.EXACTLY
+
+    const normaliseWhitespace = s => s.trim().replace(/s+/g, ' ')
+    if (normaliseWhitespace(text1) === normaliseWhitespace(text2))
```

Comment:
I'm surprised this check is needed and you don't get something like 0.99 in this case